### PR TITLE
@sweir27: Merge desktop & mobile setup code

### PR DESCRIPTION
**Pull request by craigspaeth**
Solves https://github.com/artsy/force/issues/1141

This merges desktop/lib/setup.coffee and mobile/lib/setup.coffee files and moves it up to the root along with a bunch of middleware that was duplicated between them. Sub-apps are now mounted in the root apps of desktop and mobile, and all of that middleware is in lib/setup.js. This also removes the need for that `isResponsive` silliness because we just direct mobile user agents to the mobile app and let the request fall through to desktop if it's not supported on the mobile side.

### CODE REVIEW

File: lib/setup.js [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    const ua = req.get('user-agent')
+    const isPhone = Boolean(
+      (ua.match(/iPhone/i) && !ua.match(/iPad/i)) ||
+      (ua.match(/Android/i) && ua.match(/Mobile/i)) ||
+      (ua.match(/Windows Phone/i)) ||
+      (ua.match(/BB10/i)) ||
+      (ua.match(/BlackBerry/i))
+    )
+    if (isPhone) require('../mobile')(req, res, next)
+    else next()
```
Comment (craigspaeth) [inline lib/setup.js]: This is where it falls through to desktop if not a mobile UA

File: desktop/components/fair_layout/templates/head.jade [hunk 2]
```diff
@@ -1,59 +1,67 @@
...
+
+if fair.get('location')
+  meta( property="og:location", content=fair.get('location').raw )
+
 if profile.iconImageUrl()
   meta( property="og:image", content=profile.iconImageUrl() )
+
 if sd.INCLUDE_ESCAPED_FRAGMENT
   meta( name='fragment', content='!' )
-if sd.IS_RESPONSIVE && (userAgent && !userAgent.match('iPad'))
```
Comment (craigspaeth) [inline desktop/components/fair_layout/templates/head.jade]: Sorry for some of the noise in this PR. I had to fix these `viewport` meta tags that were using `sd.IS_RESPONSIVE` which is now removed. In the process I organized the rest of the template by grouping similar tags together and sorting alphabetically.

File: desktop/components/main_layout/templates/error.jade [hunk 3]
```diff
@@ -1,27 +1,26 @@
-extends ./index
+doctype html
```
Comment (craigspaeth) [inline desktop/components/main_layout/templates/error.jade]: I was starting to address this issue https://github.com/artsy/force/issues/1148 but I should probably leave that for another PR—will update.

File: desktop/index.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
+const app = module.exports = require('express')()
```
Comment (craigspaeth) [inline desktop/index.js]: Like I mentioned in the description, these root express apps for desktop and mobile sides are now where all of their sub-apps are mounted.

File: desktop/apps/categories2/index.coffee [hunk 5]
```diff
@@ -1,6 +1,6 @@
 express = require 'express'
 routes = require './routes'
-adminOnly = require '../../lib/middleware/admin_only'
+adminOnly = require '../../lib/admin_only'
```
Comment (craigspaeth) [inline desktop/apps/categories2/index.coffee]: `admin_only` was the only middleware that was used not globally in `desktop`, so I decided to move all of the middleware up to ./lib/middleware and get rid of the folder.

File: index.js [hunk 6]
```diff
@@ -1,96 +1,23 @@
 require('coffee-script/register')
-
-// Use bluebird for performance
-global.Promise = require('bluebird')
```
Comment (craigspaeth) [inline index.js]: I thought I had kept this in, but doesn't appear to be in the diff—updating.

File: index.js [hunk 7]
```diff
@@ -1,96 +1,23 @@
...
-
-// Merge sharify data
```
Comment (craigspaeth) [inline index.js]: This file is cleaned up quite a bit now that we don't have to do weird things to merge code from desktop and mobile.

File: desktop/lib/middleware/hsts.coffee [hunk 8]
```diff
@@ -1,14 +0,0 @@
-#
-# HSTS allows for a more effective implementation of TLS by ensuring
```
Comment (craigspaeth) [inline desktop/lib/middleware/hsts.coffee]: This diff makes it look like files are being deleted, but it's just middleware being de-duplicated between desktop/mobile as you can see from [this line](https://github.com/artsy/force/pull/1197/files#diff-59f3431023761d7afa724572082feebb). Tests have also been moved up to ./test/lib.

File: lib/middleware/locals.coffee [hunk 9]
```diff
@@ -38,5 +39,14 @@ module.exports = (req, res, next) ->
   res.locals.sd.REFLECTION = req.headers?['user-agent']?.match('PhantomJS')?
   res.locals.sd.REQUEST_TIMESTAMP = Date.now()
   res.locals.sd.NOTIFICATION_COUNT = req.cookies?['notification-count']
+  res.locals.sd.USER_AGENT = res.locals.userAgent = req.get('user-agent')
+  ua = req.get('user-agent')
+  res.locals.sd.IS_MOBILE = Boolean(
```
Comment (craigspaeth) [inline lib/middleware/locals.coffee]: This `locals` middleware is merged from desktop/mobile, and this line where we set `sd.IS_MOBILE` is where we determine [whether to route to the mobile app or desktop here](https://github.com/artsy/force/pull/1197/files#diff-f5d8c0f33aa88aa6e119a0cdd5f1c40eR210).

File: lib/setup.js [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+    if (options.headers == null) { options.headers = {} }
+    options.headers['X-XAPP-TOKEN'] = artsyXapp.token || ''
+    return superSync(method, model, options)
+  }
+
+  // Inject sharify data before any app code
+  app.use(sharify)
+
+  // Add stubbed API servers and a stubbed user for testing
+  // TODO: https://github.com/artsy/force/issues/1149
```
Comment (craigspaeth) [inline lib/setup.js]: I actually completed this TODO, so I need to remove this part where we mount fake gravity/positron servers for testing. Updating...

**Comment by craigspaeth** [comment]
Looks like I have a handful of things to clean up, closing for now. Sorry for the huge diff—a lot of it is just moving files around though.

**Comment by sweir27** [comment]
Nice cleanup here! Merging so we can test in staging :D

File: desktop/test/lib/setup.coffee [hunk 11]
```diff
@@ -1,136 +0,0 @@
...
-        NODE_ENV: 'test'
-        sharify: sinon.stub().returns((req, res, next) -> next())
-      @app = express()
-      sinon.spy @app, 'use'
-      sinon.spy @app, 'get'
-      sinon.spy @app, 'set'
-      @setup @app
-
-    it 'mounts test middleware', ->
-      @app.use.args[6][0].should.equal '/__gravity'
```
Comment (kanaabe) [inline desktop/test/lib/setup.coffee]: I think this test file got dropped in this PR! I can add it back 😄  -- took me a long time to write these so I would like to see them in there unless you think we don't need them @craigspaeth
